### PR TITLE
Policies

### PR DESCRIPTION
**Pull request by ChiaraOggeri**
# Description
Created a policies folder that contains a OPA policy to verify the correct creation (or update) of a new instance.
In particular it checks if the instance referres to an existing template in the correct namespace.
Fixes # (issue)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: policies/verify-instances-templates-reference/test/rego-rule.rego [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+            var == {}
+            msg := sprintf("Namespace %v doesn't contain any template",[ns])          
+          }
+          
+          violation[{"msg":msg, "details":{"missing_tempalte": [missing]}}] {
+            provided := { input.review.object.spec["template.crownlabs.polito.it/TemplateRef"].name}
+            required := {data.inventory.namespace[ns]["crownlabs.polito.it/v1alpha2"]["Template"][_].metadata.name}
+            missing:=provided-required
+            count(missing)>0
+            msg:=sprintf("template errato %v",[missing])
```
Comment (giorio94) [inline policies/verify-instances-templates-reference/test/rego-rule.rego:22]: The error message shouldn't be in Italian

File: policies/verify-instances-templates-reference/constraintTemplate.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
+---
+apiVersion: templates.gatekeeper.sh/v1beta1
+kind: ConstraintTemplate
+metadata: 
+  name: crownlabsconstraint
```
Comment (giorio94) [inline policies/verify-instances-templates-reference/constraintTemplate.yaml:5]:
```suggestion
  name: crownlabs-instance-template-reference
```

I would suggest the same for the CRD name (in the correct case)

**Review by giorio94 — COMMENTED** [review]
First of all, thanks @ChiaraOggeri and @sofymunari for this. Looks very interesting to me.

A couple of generic comments (don't know if this PR is already ready for review or not):
* Please, use `opa fmt` to format the rego files.
* The rego policy is replicated both in its own file (for testing) and in the gatekeeper constraint. This will soon lead to inconsistencies, since one file is updated and the other not. I would suggest trying to use [konstraint](https://github.com/plexsystems/konstraint) to generate the gatekeeper constraints from the base rego policies.

File: policies/verify-instances-templates-reference/README.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+## Policies
+
+**Severity:** Violation
+
+**Resources:** crownlabs.polito.it/Instance[enter link description here](../../operators/deploy/crds/crownlabs.polito.it_instances.yaml)
+
+
+
+### Rego
```
Comment (palexster) [inline policies/verify-instances-templates-reference/README.md:23]: I personally would avoid to have code in documentation... The risk is that it easily become not updated with the code in the repo.

File: policies/verify-tenant-patch/policies/rego_rule_test.rego [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+						"fieldsV1": {"f:spec": {".": {}, "f:ID": {}, "f:createSandbox": {}, "f:name": {}, "f:publicKeys": {}, "f:surname": {}, "f:workspaces": {}}}, "manager": "kubectl-create",
+					},
+					{
+						"fieldsType": "FieldsV1", "fieldsV1": {"f:spec": {"f:email": {}}},
+						"manager": "kubectl-patch",
+						"operation": "Update",
+						"time": "2020-12-20T11:30:12Z",
+						"apiVersion": "crownlabs.polito.it/v1alpha1",
+					},
+				],
```
Comment (giorio94) [inline policies/verify-tenant-patch/policies/rego_rule_test.rego:87]: You can drop the managedFields, to shorten this file. The same for all occurrences

File: policies/verify-instances-templates-reference/README.md [hunk 5]
```diff
@@ -0,0 +1,17 @@
+VERIFY INSTANCES TEMPLATES REFERENCE
+=====================================
+## GOALS 
+This policy verifies that an instance  refers to an existing template in the correct namespace when it is created or updated.
+
+## TESTS
+Tests are available in folder [policies](./policies).
+
+## HOW TO RUN
```
Comment (giorio94) [inline policies/verify-instances-templates-reference/README.md:9]:
```suggestion
## HOW TO DEPLOY
```

**Review by giorio94 — COMMENTED** [review]
Looks in a pretty good shape. A couple of minor additional comments inline.

I just have some concerns about the new policy, but I still have to understand which would be the best solution from the design point of view. We can discuss about that on slack

**Review by palexster — APPROVED** [review]

**Review by giorio94 — APPROVED** [review]

**Pull request merged**